akka-stream-tests/src/test/scala/akka/stream/io/TcpSpec.scala

```diff
@@ -470,6 +474,38 @@ class TcpSpec extends AkkaSpec("akka.io.tcp.windows-connection-abort-workaround-
       Await.result(binding4.unbind(), 1.second)
     }
 
+    "not shut down connections after the connection stream cancelled" in assertAllStagesStopped {
+      val address = temporaryServerAddress()
+      Tcp().bind(address.getHostName, address.getPort).take(1).runForeach(_.flow.join(Flow[ByteString]).run())
+
+      val total = Source(immutable.Iterable.fill(1000)(ByteString(0)))
+        .via(Tcp().outgoingConnection(address))
+        .runFold(0)(_ + _.size)
+
+      Await.result(total, 3.seconds) should ===(1000)
+    }
+
+    "shut down properly even if some accepted connection Flows have not been subscribed to" in assertAllStagesStopped {
+      val address = temporaryServerAddress()
+      val takeTwoAndDropSecond = Flow[IncomingConnection].grouped(2).take(1).map(_.head)
+      Tcp().bind(address.getHostName, address.getPort)
+        .via(takeTwoAndDropSecond)
+        .runForeach(_.flow.join(Flow[ByteString]).run())
+
+      val folder = Source(immutable.Iterable.fill(1000)(ByteString(0)))
+        .via(Tcp().outgoingConnection(address))
+        .toMat(Sink.fold(0)(_ + _.size))(Keep.right)
+
+      val total = folder.run()
+      val rejected = folder.run()
+
+      Await.result(total, 3.seconds) should ===(1000)
+
+      a[StreamTcpException] should be thrownBy {
+        Await.result(rejected, 5.seconds) should ===(1000)
+      }
+    }
+
```

Comment:
No need to, because of the take(1) on the bind stream, it is cancelled anyway. If there is a leak because of the failed test, then the actor system shutdown will take care of it at the end of the Spec.